File: web/src/lib/requests.ts
import { z } from 'zod'
import { queryOptions, type QueryClient } from '@tanstack/react-query'
import { authKey, type AuthState } from './auth'
import { request } from './request'
import { providers } from './accounts'

export function cacheHitRate(input: number | null, cached: number | null) {
  // Missing or inconsistent usage must not look like a cache miss or a 100% hit.
  if (
    input === null ||
    cached === null ||
    input <= 0 ||
    cached < 0 ||
    cached > input
  )
    return null
  return cached / input
}

export const outcomes = [
  'success',
  'incomplete',
  'error',
  'canceled',
  'rejected',
] as const
export const outcomeKeys = {
  success: 'requestSuccess',
  incomplete: 'requestIncomplete',
  error: 'requestFailed',
  canceled: 'requestCanceled',
  rejected: 'requestRejected',
} as const
const tokens = z.number().int().nonnegative().nullable()
const recordSchema = z.object({
  id: z.number().int().positive(),
  user_id: z.number().int(),
  key_id: z.number().int(),
  group_id: z.number().int(),
  account_id: z.string(),
  provider: z.union([z.enum(providers), z.literal('')]),
  model: z.string(),
  transport: z.enum(['http', 'websocket']),
  operation: z.enum(['responses', 'chat', 'compact', 'messages', 'gemini']),
  started_at: z.number().int(),
  duration_ms: z.number().int().nonnegative(),
  request_id: z.string().max(64),
  first_token_ms: tokens,
  outcome: z.enum(outcomes),
  error_code: z.string(),
  upstream_status: z.number().int().nullable(),
  input_tokens: tokens,
  output_tokens: tokens,
  cached_tokens: tokens,
  username: z.string(),
  key_name: z.string(),
  group_name: z.string(),
  account_name: z.string(),
})
export type RequestRecord = z.infer<typeof recordSchema>
export type RequestFilters = {
  account_id?: string
  outcome?: string
  model?: string
  request_id?: string
  from?: number
  until?: number
  key_id?: number
  user_id?: number
}

function currentScope(
  client: QueryClient,
  userID: number | undefined,
  scope: 'personal' | 'all',
) {
  const user = client.getQueryData<AuthState>(authKey)?.user
  return (
    userID !== undefined &&
    user?.id === userID &&
    (scope === 'personal' || user.role === 'admin')
  )
}

export function requestOptions(
  client: QueryClient,
  userID: number | undefined,
  scope: 'personal' | 'all',
  cursor: number,
  filters: RequestFilters,
) {
  const query = new URLSearchParams({ cursor: String(cursor) })
  for (const [key, value] of Object.entries(filters)) {
    if (scope === 'personal' && (key === 'account_id' || key === 'user_id'))
      continue
    if (value !== undefined && value !== '') query.set(key, String(value))
  }
  return queryOptions({
    queryKey: ['request-history', userID, scope, cursor, filters],
    enabled: () => currentScope(client, userID, scope),
    queryFn: ({ signal }: { signal: AbortSignal }) =>
      request(
        `${scope === 'personal' ? '/api/me/requests' : '/api/requests'}?${query}`,
        z.object({
          requests: z.array(recordSchema).max(50),
          next_cursor: z.number().int().nonnegative(),
        }),
        { signal },
      ),
  })
}

export function requestCallerOptions(
  client: QueryClient,
  userID: number,
  scope: 'personal' | 'all',
) {
  return queryOptions({
    queryKey: ['request-callers', userID, scope],
    enabled: () => currentScope(client, userID, scope),
    queryFn: ({ signal }: { signal: AbortSignal }) =>
      request(
        `${scope === 'personal' ? '/api/me/requests' : '/api/requests'}/filters`,
        z.object({
          callers: z
            .array(
              z.object({
                user_id: z.number().int(),
                key_id: z.number().int(),
                username: z.string(),
                key_name: z.string(),
              }),
            )
            .max(5000),
        }),
        { signal },
      ),
    staleTime: 30000,
  })
}

File: internal/gateway/requests.go
package gateway

import (
	"context"
	"time"

	"github.com/murongg/SubLane/internal/accounts"
	"github.com/murongg/SubLane/internal/storage/db"
)

type RequestRecord struct {
	ID             int64  `json:"id"`
	UserID         int64  `json:"user_id"`
	KeyID          int64  `json:"key_id"`
	GroupID        int64  `json:"group_id"`
	AccountID      string `json:"account_id"`
	Provider       string `json:"provider"`
	Model          string `json:"model"`
	Transport      string `json:"transport"`
	Operation      string `json:"operation"`
	StartedAt      int64  `json:"started_at"`
	DurationMs     int64  `json:"duration_ms"`
	Outcome        string `json:"outcome"`
	ErrorCode      string `json:"error_code"`
	UpstreamStatus *int64 `json:"upstream_status"`
	InputTokens    *int64 `json:"input_tokens"`
	OutputTokens   *int64 `json:"output_tokens"`
	CachedTokens   *int64 `json:"cached_tokens"`
	RequestID      string `json:"request_id"`
	FirstTokenMs   *int64 `json:"first_token_ms"`
	Username       string `json:"username"`
	KeyName        string `json:"key_name"`
	GroupName      string `json:"group_name"`
	AccountName    string `json:"account_name"`
}
type RequestPage struct {
	Requests   []RequestRecord `json:"requests"`
	NextCursor int64           `json:"next_cursor"`
}

type RequestFilter struct {
	Cursor                               int64
	AccountID, Outcome, Model, RequestID string
	From, Until, KeyID, MemberID         int64
}

func (s *Service) Requests(ctx context.Context, filter RequestFilter) (RequestPage, error) {
	return s.requests(ctx, 0, filter)
}
func (s *Service) UserRequests(ctx context.Context, userID int64, filter RequestFilter) (RequestPage, error) {
	if userID <= 0 {
		return RequestPage{}, accounts.ErrInput
	}
	// Caller-controlled operator filters can never replace the session's ownership scope.
	filter.MemberID = 0
	filter.AccountID = ""
	return s.requests(ctx, userID, filter)
}
func (s *Service) requests(ctx context.Context, userID int64, f RequestFilter) (RequestPage, error) {
	page := RequestPage{Requests: []RequestRecord{}}
	if f.Cursor < 0 || len(f.AccountID) > 64 || f.KeyID < 0 || f.MemberID < 0 || f.From < 0 || f.Until < 0 || (f.Until != 0 && f.From >= f.Until) || (f.Model != "" && !safeModel.MatchString(f.Model)) || (f.RequestID != "" && !safeRequestID.MatchString(f.RequestID)) {
		return page, accounts.ErrInput
	}
	switch f.Outcome {
	case "", "success", "incomplete", "error", "canceled", "rejected":
	default:
		return page, accounts.ErrInput
	}
	since := s.now().Add(-7 * 24 * time.Hour).Unix()
	if err := s.queries.PruneRequests(ctx, since); err != nil {
		return page, err
	}
	rows, err := s.queries.ListRequests(ctx, db.ListRequestsParams{TenantID: s.tenantID, UserID: userID, Cursor: f.Cursor, AccountID: f.AccountID, Outcome: f.Outcome, Since: max(since, f.From), UntilTime: f.Until, MemberID: f.MemberID, KeyID: f.KeyID, Model: f.Model, RequestID: f.RequestID})
	if err != nil {
		return page, err
	}
	for _, row := range rows {
		record := RequestRecord(row)
		if userID != 0 {
			record.AccountID, record.AccountName = "", ""
		}
		page.Requests = append(page.Requests, record)
	}
	if len(page.Requests) > 50 {
		page.Requests = page.Requests[:50]
		page.NextCursor = page.Requests[49].ID
	}
	return page, nil
}

type RequestCaller struct {
	UserID   int64  `json:"user_id"`
	KeyID    int64  `json:"key_id"`
	Username string `json:"username"`
	KeyName  string `json:"key_name"`
}

func (s *Service) RequestCallers(ctx context.Context, userID int64) ([]RequestCaller, error) {
	rows, err := s.queries.ListRequestCallers(ctx, db.ListRequestCallersParams{TenantID: s.tenantID, UserID: userID, Since: s.now().Add(-7 * 24 * time.Hour).Unix()})
	if err != nil {
		return nil, err
	}
	result := make([]RequestCaller, 0, len(rows))
	for _, row := range rows {
		result = append(result, RequestCaller(row))
	}
	return result, nil
}
